sdk/search/Azure.Search.Documents/src/Batching/ProducerBarrierSlim.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Search.Documents.Batching
+{
+    /// <summary>
+    /// While most writers of a Channel only want to push work and return,
+    /// occasionally we'll have writers like Flush that want to block until
+    /// the reader does something with that work.  ProducerBarrierSlim allows
+    /// them to wait and be released after the work has completed.
+    /// </summary>
+    internal class ProducerBarrierSlim : IDisposable
+    {
+        /// <summary>
+        /// Semaphore to enforce the barrier.
+        /// </summary>
+        private SemaphoreSlim _semaphore = new SemaphoreSlim(0);
+
+        /// <summary>
+        /// The number of tasks waiting on the submission semaphore.  This
+        /// determines how many times we release the semaphore.
+        /// </summary>
+        internal volatile int _waiters = 0;
+
+        /// <summary>
+        /// Wait for the barrier to be lifted.
+        /// </summary>
+        /// <param name="async">Whether to call sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>
+        /// A Task that will wait until the barrier is lifted.
+        /// </returns>
+        public async Task WaitAsync(bool async, CancellationToken cancellationToken)
+        {
+            if (_semaphore == null)
+            {
+                throw new ObjectDisposedException(nameof(ProducerBarrierSlim));
+            }
+
+            Interlocked.Increment(ref _waiters);
+            try
+            {
+                if (async)
+                {
+                    await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    _semaphore.Wait(cancellationToken);
+                }
+            }
+            finally
+            {
+                Interlocked.Decrement(ref _waiters);
+            }
+        }
+
+        /// <summary>
+        /// Lift the barrier.
+        /// </summary>
+        public void Release()
```

Comment:
Can release be called from parallel threads? If so this implementation is racy. Two threads can get the same value in `waiting = _waiters` and call `_semaphore.Release` too many times.

Reply:
No, it's only ever called from the single consumer.

Reply:
Why is the field volatile then?

Reply:
Wait gets called by all the producer threads.  Release gets called only by the single consumer.

Reply:
`SemaphoreSlim.Release` should be a full memory barrier in itself.

Reply:
Can WaitAsync happen while the Release is running?

Reply:
Yes, it could.  Per your other comment I'll probably need another guard around this anyway.